File: ios/Yappy/Features/Group/GroupModeration.swift
import SwiftUI

/// Moderation surfaces that are too big to live inside the settings scroll:
/// the ban list, and invite-link management.

// ── Shared permission arithmetic ─────────────────────────────────────────────

/// The conversation-wide permission floor, as the server computes it.
///
/// Duplicated from `packages/shared/src/permissions.ts` for the same reason
/// `Perm` in GroupSettingsScreen is: these bits are part of the wire format and
/// cannot move without a coordinated release. Only the two floors the UI can
/// actually set are mirrored — the full table stays server-side.
enum BaseFloor {
    private static func bit(_ index: Int64) -> Int64 { Int64(1) << index }

    /// What an ordinary group gives everyone: read, write, react, call, invite.
    static let member: Int64 = {
        let send = bit(2) | bit(3) | bit(4) | bit(5) | bit(6) | bit(7) | bit(8) | bit(10)
        let own = bit(11) | bit(12)
        return bit(0) | bit(1) | send | own | bit(20) | bit(21) | bit(23) | bit(30)
    }()

    /// Announcement mode: everyone may read and react, nobody may post. Roles
    /// hand posting back to the people who should still have it, which is why
    /// this works as a *floor* rather than as a lock.
    static let announcement: Int64 = bit(0) | bit(1) | bit(8)

    /// A conversation with no explicit base inherits `member`, so nil reads as
    /// "everyone can post".
    static func isAnnouncement(_ raw: String?) -> Bool {
        guard let raw, let bits = Int64(raw) else { return false }
        return bits & bit(2) == 0
    }
}

// ── Ban list ─────────────────────────────────────────────────────────────────

/// Who has been thrown out, and the way back in.
///
/// Until this existed a ban was a one-way door: the API could set one and clear
/// one, but nothing could tell you a ban was there, so an accidental ban was
/// unrecoverable from inside the app.
struct BanListSheet: View {
    @Environment(\.neu) private var colors
    @EnvironmentObject private var container: AppContainer

    let conversationId: String

    @State private var bans: [BanEntry]?
    @State private var working: Set<String> = []

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                Text("Banned")
                    .font(YappyFont.titleMedium)
                    .foregroundStyle(colors.textPrimary)

                Text("Someone who is banned cannot rejoin, even with an invite link.")
                    .font(YappyFont.labelSmall)
                    .foregroundStyle(colors.textTertiary)
                    .padding(.top, 4)
                    .padding(.bottom, 14)

                if bans == nil {
                    NeuSpinner().frame(maxWidth: .infinity).padding(.vertical, 30)
                } else if bans?.isEmpty == true {
                    Text("Nobody is banned.")
                        .font(YappyFont.bodyMedium)
                        .foregroundStyle(colors.textTertiary)
                        .padding(.vertical, 20)
                }

                ForEach(bans ?? []) { ban in
                    row(ban)
                }
            }
            .padding(.horizontal, 16)
            .padding(.top, 20)
            .padding(.bottom, 28)
        }
        .task { bans = (try? await container.repo.bans(conversationId).bans) ?? [] }
    }

    private func row(_ ban: BanEntry) -> some View {
        HStack(spacing: 12) {
            Avatar(url: ban.user.avatarUrl, name: ban.user.label, id: ban.user.id, size: 38)

            VStack(alignment: .leading, spacing: 1) {
                Text(ban.user.label)
                    .font(YappyFont.bodyLarge)
                    .foregroundStyle(colors.textPrimary)
                if let reason = ban.reason, !reason.isEmpty {
                    Text(reason)
                        .font(YappyFont.labelSmall)
                        .foregroundStyle(colors.textTertiary)
                        .lineLimit(1)
                } else if let when = ban.createdAt {
                    Text(YappyTime.relative(when))
                        .font(YappyFont.labelSmall)
                        .foregroundStyle(colors.textTertiary)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            Text(working.contains(ban.id) ? "…" : "Unban")
                .font(YappyFont.labelMedium)
                .foregroundStyle(colors.accent)
                .softTap(enabled: !working.contains(ban.id)) {
                    working.insert(ban.id)
                    Task {
                        try? await container.repo.unban(conversationId, userId: ban.user.id)
                        bans?.removeAll { $0.id == ban.id }
                        working.remove(ban.id)
                    }
                }
        }
        .padding(.vertical, 8)
    }
}

// ── Invite links ─────────────────────────────────────────────────────────────

private let expiryChoices: [(String, Int?)] = [
    ("Never", nil),
    ("1 day", 86_400),
    ("7 days", 604_800),
    ("30 days", 2_592_000),
]

private let usesChoices: [(String, Int)] = [
    ("Unlimited", 0),
    ("1 use", 1),
    ("5 uses", 5),
    ("25 uses", 25),
]

/// Create, share and revoke invite links.
///
/// A group with one permanent unlimited link has no way to un-share it once it
/// leaks, which is the actual failure mode for a public group — hence revoke,
/// and hence expiry as the default-shaped option rather than a buried one.
struct InviteManagerSheet: View {
    @Environment(\.neu) private var colors
    @EnvironmentObject private var container: AppContainer

    let conversationId: String

    @State private var invites: [Invite]?
    @State private var expiry: Int?
    @State private var uses = 0
    @State private var busy = false
    @State private var copiedCode: String?

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                Text("Invite links")
                    .font(YappyFont.titleMedium)
                    .foregroundStyle(colors.textPrimary)
                    .padding(.bottom, 14)

                creator

                if invites == nil {
                    NeuSpinner().frame(maxWidth: .infinity).padding(.vertical, 26)
                } else if invites?.isEmpty == true {
                    Text("No links yet.")
                        .font(YappyFont.bodyMedium)
                        .foregroundStyle(colors.textTertiary)
                        .padding(.top, 20)
                }

                ForEach(invites ?? []) { invite in
                    row(invite).padding(.top, 10)
                }
            }
            .padding(.horizontal, 16)
            .padding(.top, 20)
            .padding(.bottom, 30)
        }
        .task { invites = (try? await container.repo.invites(conversationId).invites) ?? [] }
    }

    private var creator: some View {
        NeuSurface(radius: Neu.cornerMedium, contentPadding: 14) {
            VStack(alignment: .leading, spacing: 12) {
                chips("Expires", expiryChoices.map(\.0), selected: expiryChoices.firstIndex { $0.1 == expiry } ?? 0) {
                    expiry = expiryChoices[$0].1
                }
                chips("Uses", usesChoices.map(\.0), selected: usesChoices.firstIndex { $0.1 == uses } ?? 0) {
                    uses = usesChoices[$0].1
                }

                NeuButton(enabled: !busy, accent: true) {
                    busy = true
                    Task {
                        if let created = try? await container.repo.createInvite(
                            conversationId, maxUses: uses, expiresInSeconds: expiry
                        ).invite {
                            invites?.insert(created, at: 0)
                        }
                        busy = false
                    }
                } content: {
                    Text(busy ? "Creating…" : "Create link")
                        .font(YappyFont.labelLarge)
                        .foregroundStyle(colors.onAccent)
                }
            }
        }
    }

    private func chips(
        _ label: String,
        _ options: [String],
        selected: Int,
        onPick: @escaping (Int) -> Void
    ) -> some View {
        VStack(alignment: .leading, spacing: 7) {
            Text(label)
                .font(YappyFont.labelSmall)
                .foregroundStyle(colors.textTertiary)

            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 8) {
                    ForEach(Array(options.enumerated()), id: \.offset) { index, option in
                        NeuChip(label: option, selected: index == selected) { onPick(index) }
                    }
                }
            }
        }
    }

    private func row(_ invite: Invite) -> some View {
        NeuSurface(radius: Neu.cornerMedium, contentPadding: 12) {
            VStack(alignment: .leading, spacing: 6) {
                HStack(spacing: 8) {
                    Image(systemName: "link")
                        .font(.system(size: 13))
                        .foregroundStyle(colors.accent)
                    Text(invite.url)
                        .font(YappyFont.bodyMedium)
                        .foregroundStyle(colors.textSecondary)
                        .lineLimit(1)
                        .frame(maxWidth: .infinity, alignment: .leading)
                    Text(copiedCode == invite.code ? "copied" : "copy")
                        .font(YappyFont.labelSmall)
                        .foregroundStyle(colors.accent)
                        .softTap {
                            UIPasteboard.general.string = invite.url
                            copiedCode = invite.code
                        }
                }

                HStack(spacing: 8) {
                    Text(describe(invite))
                        .font(YappyFont.labelSmall)
                        .foregroundStyle(colors.textTertiary)
                    Spacer(minLength: 0)
                    Text("revoke")
                        .font(YappyFont.labelSmall)
                        .foregroundStyle(colors.danger)
                        .softTap {
                            Task {
                                try? await container.repo.revokeInvite(conversationId, code: invite.code)
                                invites?.removeAll { $0.code == invite.code }
                            }
                        }
                }
            }
        }
    }

    private func describe(_ invite: Invite) -> String {
        var parts: [String] = []
        parts.append(invite.maxUses == 0 ? "\(invite.uses) uses" : "\(invite.uses)/\(invite.maxUses) used")
        if let expires = invite.expiresAt { parts.append("expires \(YappyTime.relative(expires))") }
        return parts.joined(separator: " · ")
    }
}

/// Pick a bot to add to this group.
///
/// Adding goes through the ordinary add-members call rather than anything
/// bot-specific: a bot is a user row, so it lands with the same permission
/// check and the same "X added Y" system message a person would. The group can
/// see it arrive, which for something that reads every message is the point.
struct BotPickerSheet: View {
    @Environment(\.neu) private var colors
    @EnvironmentObject private var container: AppContainer

    let conversationId: String

    @State private var bots: [DirectoryBot]?
    @State private var adding: String?
    @State private var added: Set<String> = []
    @State private var error: String?

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                Text("Add a bot")
                    .font(YappyFont.titleMedium)
                    .foregroundStyle(colors.textPrimary)

                Text("It will be able to read this group's messages and post in it.")
                    .font(YappyFont.labelSmall)
                    .foregroundStyle(colors.textTertiary)
                    .padding(.top, 4)
                    .padding(.bottom, 14)

                if bots == nil {
                    NeuSpinner().frame(maxWidth: .infinity).padding(.vertical, 30)
                } else if bots?.isEmpty == true {
                    Text("No public bots yet. Build one in the developer portal and mark it public.")
                        .font(YappyFont.bodyMedium)
                        .foregroundStyle(colors.textTertiary)
                        .padding(.vertical, 20)
                }

                ForEach(bots ?? []) { bot in
                    row(bot).padding(.bottom, 10)
                }

                if let error {
                    Text(error)
                        .font(YappyFont.bodyMedium)
                        .foregroundStyle(colors.danger)
                        .padding(.top, 12)
                }
            }
            .padding(.horizontal, 16)
            .padding(.top, 20)
            .padding(.bottom, 28)
        }
        .task { bots = (try? await container.repo.botDirectory().bots) ?? [] }
    }

    private func row(_ bot: DirectoryBot) -> some View {
        let isAdded = added.contains(bot.botUserId)
        let blurb = bot.description?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let subtitle = blurb.isEmpty
            ? (bot.commandCount > 0 ? "\(bot.commandCount) commands" : "Bot")
            : blurb

        return NeuSurface(radius: Neu.cornerMedium, contentPadding: 14) {
            HStack(spacing: 12) {
                Avatar(url: bot.user?.avatarUrl, name: bot.name, id: bot.botUserId, size: 40)

                VStack(alignment: .leading, spacing: 2) {
                    Text(bot.name)
                        .font(YappyFont.bodyLarge)
                        .foregroundStyle(colors.textPrimary)
                        .lineLimit(1)
                    Text(subtitle)
                        .font(YappyFont.bodyMedium)
                        .foregroundStyle(colors.textTertiary)
                        .lineLimit(2)
                }

                Spacer(minLength: 0)

                if isAdded {
                    Text("Added")
                        .font(YappyFont.labelLarge)
                        .foregroundStyle(colors.success)
                } else {
                    NeuButton(enabled: adding == nil, accent: true) {
                        add(bot)
                    } content: {
                        if adding == bot.botUserId {
                            NeuSpinner(tint: colors.onAccent)
                        } else {
                            Text("Add")
                                .font(YappyFont.labelLarge)
                                .foregroundStyle(colors.onAccent)
                        }
                    }
                    .frame(width: 96)
                }
            }
        }
    }

    private func add(_ bot: DirectoryBot) {
        guard adding == nil else { return }
        adding = bot.botUserId
        error = nil
        Task {
            do {
                try await container.repo.addMembers(conversationId, userIds: [bot.botUserId])
                added.insert(bot.botUserId)
            } catch {
                self.error = "Could not add \(bot.name)."
            }
            adding = nil
        }
    }
}

File: android/app/src/main/java/gg/yappy/app/ui/group/GroupModeration.kt
package gg.yappy.app.ui.group

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import gg.yappy.app.LocalContainer
import gg.yappy.app.data.BanEntry
import gg.yappy.app.data.Invite
import gg.yappy.app.ui.components.Avatar
import gg.yappy.app.ui.components.NeuButton
import gg.yappy.app.ui.components.NeuChip
import gg.yappy.app.ui.components.softClickable
import gg.yappy.app.ui.theme.neuColors
import gg.yappy.app.ui.util.relativeTime
import kotlinx.coroutines.launch
import gg.yappy.app.data.RoleEntry

/**
 * Moderation surfaces that are too big to live inside the settings scroll: the
 * ban list, and invite-link management.
 */

// ── Permission arithmetic ────────────────────────────────────────────────────

/**
 * The conversation-wide permission floor, as the server computes it.
 *
 * Duplicated from `packages/shared/src/permissions.ts` because these bits are
 * part of the wire format and cannot move without a coordinated release. Only
 * the two floors the UI can actually set are mirrored — the full table stays
 * server-side.
 */
object BaseFloor {
    private fun bit(index: Int): Long = 1L shl index

    /** What an ordinary group gives everyone: read, write, react, call, invite. */
    val MEMBER: Long = run {
        val send = bit(2) or bit(3) or bit(4) or bit(5) or bit(6) or bit(7) or bit(8) or bit(10)
        val own = bit(11) or bit(12)
        bit(0) or bit(1) or send or own or bit(20) or bit(21) or bit(23) or bit(30)
    }

    /**
     * Announcement mode: everyone may read and react, nobody may post. Roles
     * hand posting back to the people who should still have it, which is why
     * this works as a *floor* rather than as a lock.
     */
    val ANNOUNCEMENT: Long = bit(0) or bit(1) or bit(8)

    /**
     * A conversation with no explicit base inherits [MEMBER], so null reads as
     * "everyone can post".
     */
    fun isAnnouncement(raw: String?): Boolean {
        val bits = raw?.toLongOrNull() ?: return false
        return bits and bit(2) == 0L
    }
}

// ── Ban list ─────────────────────────────────────────────────────────────────

/**
 * Who has been thrown out, and the way back in.
 *
 * Until this existed a ban was a one-way door: the API could set one and clear
 * one, but nothing on Android could tell you a ban was there, so an accidental
 * ban was unrecoverable from inside the app.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun BanListSheet(conversationId: String, onDismiss: () -> Unit) {
    val container = LocalContainer.current
    val colors = neuColors
    val scope = rememberCoroutineScope()

    /** Null while loading, so the empty state does not flash before the answer. */
    var bans by remember { mutableStateOf<List<BanEntry>?>(null) }
    var working by remember { mutableStateOf<Set<String>>(emptySet()) }

    LaunchedEffect(conversationId) {
        bans = runCatching { container.repo.bans(conversationId).bans }.getOrDefault(emptyList())
    }

    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = colors.surface,
        contentColor = colors.textPrimary,
    ) {
        Column(Modifier.fillMaxWidth().padding(horizontal = 20.dp).padding(bottom = 28.dp)) {
            Text("Banned", style = MaterialTheme.typography.titleMedium, color = colors.textPrimary)
            Spacer(Modifier.padding(top = 4.dp))
            Text(
                "Someone who is banned cannot rejoin, even with an invite link.",
                style = MaterialTheme.typography.labelSmall,
                color = colors.textTertiary,
            )
            Spacer(Modifier.padding(top = 14.dp))

            val list = bans
            when {
                list == null -> Box(Modifier.fillMaxWidth().padding(vertical = 30.dp), Alignment.Center) {
                    CircularProgressIndicator(color = colors.accent)
                }

                list.isEmpty() -> Text(
                    "Nobody is banned.",
                    style = MaterialTheme.typography.bodyMedium,
                    color = colors.textTertiary,
                    modifier = Modifier.padding(vertical = 20.dp),
                )

                else -> list.forEach { ban ->
                    Row(
                        Modifier.fillMaxWidth().padding(vertical = 8.dp),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Avatar(ban.user.avatarUrl, ban.user.label, ban.user.id, size = 38.dp)
                        Spacer(Modifier.width(12.dp))
                        Column(Modifier.weight(1f)) {
                            Text(
                                ban.user.label,
                                style = MaterialTheme.typography.bodyLarge,
                                color = colors.textPrimary,
                            )
                            val detail = ban.reason?.takeIf { it.isNotBlank() }
                                ?: ban.createdAt?.let(::relativeTime)
                            detail?.let {
                                Text(
                                    it,
                                    style = MaterialTheme.typography.labelSmall,
                                    color = colors.textTertiary,
                                    maxLines = 1,
                                )
                            }
                        }
                        Text(
                            if (working.contains(ban.user.id)) "…" else "Unban",
                            style = MaterialTheme.typography.labelMedium,
                            color = colors.accent,
                            modifier = Modifier.softClickable(
                                enabled = !working.contains(ban.user.id),
                            ) {
                                working = working + ban.user.id
                                scope.launch {
                                    // Dropped from the list only once the server
                                    // agrees, so a failed call leaves the row
                                    // there to try again rather than pretending.
                                    if (runCatching { container.repo.unban(conversationId, ban.user.id) }
                                            .isSuccess
                                    ) {
                                        bans = bans?.filterNot { it.user.id == ban.user.id }
                                    }
                                    working = working - ban.user.id
                                }
                            },
                        )
                    }
                }
            }
        }
    }
}

// ── Invite links ─────────────────────────────────────────────────────────────

/**
 * Existing invite links, and the way to make or revoke one.
 *
 * The Android build could only ever create a single unlimited, never-expiring
 * link — the one shape you would not want for a link posted publicly. This
 * offers the limits the API has always accepted, and lets a link that has been
 * shared too widely be taken back.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun InviteManagerSheet(conversationId: String, onDismiss: () -> Unit) {
    val container = LocalContainer.current
    val colors = neuColors
    val scope = rememberCoroutineScope()
    val clipboard = androidx.compose.ui.platform.LocalClipboardManager.current

    var invites by remember { mutableStateOf<List<Invite>?>(null) }
    var maxUses by remember { mutableStateOf(0) }
    var expiresIn by remember { mutableStateOf<Int?>(null) }
    /** Roles new links can grant. Empty in a DM, where none exist. */
    var roles by remember { mutableStateOf<List<RoleEntry>>(emptyList()) }
    var grantRoleId by remember { mutableStateOf<String?>(null) }
    var busy by remember { mutableStateOf(false) }

    LaunchedEffect(conversationId) {
        invites = runCatching { container.repo.invites(conversationId).invites }.getOrDefault(emptyList())
        roles = runCatching { container.repo.roles(conversationId).roles }.getOrDefault(emptyList())
    }

    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = colors.surface,
        contentColor = colors.textPrimary,
    ) {
        Column(Modifier.fillMaxWidth().padding(horizontal = 20.dp).padding(bottom = 28.dp)) {
            Text("Invite links", style = MaterialTheme.typography.titleMedium, color = colors.textPrimary)
            Spacer(Modifier.padding(top = 14.dp))

            Text("Uses", style = MaterialTheme.typography.labelSmall, color = colors.textTertiary)
            Spacer(Modifier.padding(top = 6.dp))
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                NeuChip("Unlimited", maxUses == 0, onClick = { maxUses = 0 })
                NeuChip("1", maxUses == 1, onClick = { maxUses = 1 })
                NeuChip("10", maxUses == 10, onClick = { maxUses = 10 })
                NeuChip("50", maxUses == 50, onClick = { maxUses = 50 })
            }

            Spacer(Modifier.padding(top = 12.dp))
            Text("Expires", style = MaterialTheme.typography.labelSmall, color = colors.textTertiary)
            Spacer(Modifier.padding(top = 6.dp))
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                NeuChip("Never", expiresIn == null, onClick = { expiresIn = null })
                NeuChip("1 hour", expiresIn == 3_600, onClick = { expiresIn = 3_600 })
                NeuChip("1 day", expiresIn == 86_400, onClick = { expiresIn = 86_400 })
                NeuChip("7 days", expiresIn == 604_800, onClick = { expiresIn = 604_800 })
            }

            /*
             * What the next link hands out. One role per link, so "who did
             * this admit as what" stays answerable — a second role is a
             * second link.
             */
            if (roles.isNotEmpty()) {
                Spacer(Modifier.padding(top = 12.dp))
                Text("Grants", style = MaterialTheme.typography.labelSmall, color = colors.textTertiary)
                Spacer(Modifier.padding(top = 6.dp))
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    NeuChip("No role", grantRoleId == null, onClick = { grantRoleId = null })
                    roles.take(3).forEach { role ->
                        NeuChip(
                            role.name,
                            grantRoleId == role.id,
                            onClick = { grantRoleId = role.id },
                        )
                    }
                }
            }

            Spacer(Modifier.padding(top = 16.dp))
            NeuButton(
                onClick = {
                    if (busy) return@NeuButton
                    busy = true
                    scope.launch {
                        runCatching { container.repo.createInvite(conversationId, maxUses, expiresIn, roleId = grantRoleId) }
                            .getOrNull()
                            ?.let { created -> invites = listOf(created.invite) + invites.orEmpty() }
                        busy = false
                    }
                },
                accent = true,
                enabled = !busy,
                modifier = Modifier.fillMaxWidth(),
            ) {
                Text(
                    if (busy) "Creating…" else "Create a link",
                    style = MaterialTheme.typography.labelLarge,
                    color = colors.onAccent,
                )
            }

            Spacer(Modifier.padding(top = 18.dp))

            val list = invites
            when {
                list == null -> Box(Modifier.fillMaxWidth().padding(vertical = 24.dp), Alignment.Center) {
                    CircularProgressIndicator(color = colors.accent)
                }

                list.isEmpty() -> Column(
                    Modifier.fillMaxWidth().padding(vertical = 20.dp),
                    horizontalAlignment = Alignment.CenterHorizontally,
                ) {
                    Text("🔗", style = MaterialTheme.typography.headlineMedium)
                    Spacer(Modifier.padding(top = 8.dp))
                    Text(
                        "No invite links yet",
                        style = MaterialTheme.typography.titleSmall,
                        color = colors.textSecondary,
                    )
                    Text(
                        "Make one above and send it to whoever belongs here.",
                        style = MaterialTheme.typography.bodySmall,
                        color = colors.textTertiary,
                    )
                }

                else -> list.forEach { invite ->
                    Row(
                        Modifier.fillMaxWidth().padding(vertical = 9.dp),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Column(
                            Modifier.weight(1f).softClickable {
                                clipboard.setText(androidx.compose.ui.text.AnnotatedString(invite.url))
                            },
                        ) {
                            Text(
                                invite.url,
                                style = MaterialTheme.typography.bodyMedium,
                                color = colors.textPrimary,
                                maxLines = 1,
                            )
                            Text(
                                buildString {
                                    append(
                                        if (invite.maxUses == 0) {
                                            (invite.role?.let { "grants ${it.name} · " } ?: "") +
                                                "${invite.uses} uses · unlimited"
                                        } else {
                                            (invite.role?.let { "grants ${it.name} · " } ?: "") +
                                                "${invite.uses}/${invite.maxUses} uses"
                                        },
                                    )
                                    invite.expiresAt?.let { append(" · expires ${relativeTime(it)}") }
                                },
                                style = MaterialTheme.typography.labelSmall,
                                color = colors.textTertiary,
                            )
                        }
                        Text(
                            "Revoke",
                            style = MaterialTheme.typography.labelMedium,
                            color = colors.danger,
                            modifier = Modifier.softClickable {
                                scope.launch {
                                    if (runCatching {
                                            container.repo.revokeInvite(conversationId, invite.code)
                                        }.isSuccess
                                    ) {
                                        invites = invites?.filterNot { it.code == invite.code }
                                    }
                                }
                            },
                        )
                    }
                }
            }
        }
    }
}
